newlib/libc/string/strcat_s.c

```diff
@@ -0,0 +1,163 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright © 2024, Synopsys Inc.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials provided
+ *    with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#define __STDC_WANT_LIB_EXT1__ 1
+#include <stdint.h>
+#include <stdbool.h>
+#include <string.h>
+#include <stdlib.h>
+
+__errno_t strcat_s(char *restrict s1, rsize_t s1max, const char *restrict s2) 
+{
+    const char *msg = "";
+    size_t s1_len = 0;
+    bool write_null = true;
+	constraint_handler_t handler = NULL;
+
+	if (s1 == NULL)
+	{
+	    msg = "strcat_s: dest is NULL";
+	    write_null = false;
+		goto handle_error;
+	}
+
+	if (s1max == 0 || s1max > RSIZE_MAX)
+	{
+	    msg = "strcat_s: dest buffer size is 0 or exceeds RSIZE_MAX";
+	    write_null = false;
+		goto handle_error;
+	}
+
+	if (s2 == NULL) 
+	{
+	    msg = "strcat_s: source is NULL";
+		goto handle_error;
+	}
+
+    /* It is a constraint violation if s1max is not large enough to contain
+     * the concatenation s2: no truncation permitted.
+     * It is also a constraint violation if the string pointed to by s2
+     * overlaps s1 in any way.
+     * The C11 Rationale says we are permitted to proceed with the copy and
+     * detect dest buffer overrun and overlapping memory blocks as a byproduct
+     * of performing the copy operation.  This is to avoid calling strlen on
+     * s2 to detect these violations prior to attempting the copy.
+     */
+	// compute chars available in s1
```

Comment:
formatting is a bit weird here, maybe just run the whole thing through `git clang-format`?

Reply:
Thank you for pointing that out. I've applied `git clang-format` to address the formatting issues and pushed the changes.